### ISSUE TITLE
License key input is wrong on X12SCZ

### ISSUE THREAD

**Issue by cryptage21** [zsrv/supermicro-product-key#1]
Hi,

I'm trying to add SFT-DCMS-SINGLE license to my X12SCZ mainboard (JSON file).

When I add it I get an exit code "149" with error message "License key input is wrong".

Steps followed : 

```
./supermicro-product-key oob encode 3cecefxxxxxx (xxxxxx is my BMC Mac address)
XXXX-XXXX-XXXX-XXXX-XXXX-XXXX
```

```
./supermicro-product-key nonjson encode --sku SFT-DCMS-SINGLE --creation-date 2023-03-01T12:00:00Z --expiration-date 1970-01-01T00:00:00Z 3cecefxxxxxx
```

I get my key AAYAAAAAAAAAAAAAAAAAAIt...==

If I try to bruteforce I get my BMC Mac address.
```
 ./supermicro-product-key nonjson bruteforce AAYAAAAAAAAAAAAAAAAAAIt...== 
searching for mac address ...
found match! mac = '3cecefxxxxxx''
```

If I try to decode, the file looks good : 
```
./supermicro-product-key nonjson decode 3cecefxxxxxx AAYAAAAAAAAAAAAAAAAAAIt...==

{
	"FormatVersion": 0,
	"SoftwareIdentifier": {
		"SKU": "SFT-DCMS-SINGLE",
		"DisplayName": "ALL",
		"ID": 6
	},
	"SoftwareVersion": "none",
	"InvoiceNumber": "none",
	"CreationDate": "2023-03-01T12:00:00Z",
	"ExpirationDate": "1970-01-01T00:00:00Z",
	"Property": null,
	"SecretData": "O....",
	"Checksum": 6X
}

```
So I've converted the output to JSON and tried to install the key : 

`sum.exe -i 192.168.1.1 -u ADMIN -p xxxxxxx -c ActivateProductKey --key '{"ProductKey":{"Node":{"LicenseID":"1","LicenseName":"SFT-DCMS-SINGLE","CreateDate":"20230301"},"Signature":"AAAAAAAAAIt...=="}}'`

Then I get (after a long time) : 

``` 
********************************<<<<<ERROR>>>>>*********************************

ExitCode                = 149
Description             = IPMI execution failed
Program Error Code      = 407.16
Error message:
        License key input is wrong.

********************************************************************************
```

I've tried with 6 as value for "LicenseID" but that's the same.

Am I missing something ? 

Thanks for your help.

**Comment by zsrv** [comment]
Hi there, unfortunately the "Signature" field in the JSON key format is an RSA signature created using an RSA private key that is probably known only to Supermicro. The corresponding RSA public key is embedded in the BMC firmware, and is used to verify the signature. Without the RSA private key, we can't generate the JSON format keys ourselves.

**Comment by cryptage21** [comment]
Thank you for the answer.
I thought that your tool integrated that private key to be able to generate the license.

So there is no way to generate the license since the motherboard only accepts JSON, too bad.

I was interested in Virtual Media in HTML5 KVM mode. 
Supermicro license is too expensive (almost price I've bought the mainboard...).

I'll use virtual HTTP media or Java KVM to install.

Thanks again for your nice work and for your help.

**Comment by cryptage21** [comment]
It won't change anything but LicenseID value has to be "2" for SFT-DCMS-SINGLE : 

```
 if (!(licenseId === "1" && licenseName === "SFT-OOB-LIC") &&
                !(licenseId === "2" && licenseName === "SFT-DCMS-SINGLE")) {
                blockUI ("red", translator.g ("LANG_MISC_LICENSE_ACTIVATION"));
```

**Comment by zsrv** [comment]
Thanks, I started working on a license verification function for the JSON format keys but had trouble figuring out what message was being hashed and signed. I think it's some combination of the BMC MAC address and the values inside the "Node" object. But verification of existing licenses would probably be the furthest we would get.

**Comment by zsrv** [comment]
JSON format key signature verification is available in release 1.1.0 now, but we won't be able to generate them.
